OSD_MESSAGES initial implementation

fiam opened the pull request:
Experimental implementation of the OSD_MESSAGES element. Allows displaying arbitrary text messages, taking a whole line of text. It's intended to display non-essential messages, so the initial idea is not enabling it by default (subject to change).

Currently used while the aircraft can't be armed to show the reason it's failing to arm.

Potential future uses include:

- Failsafe stages
- RTH stages
- WPT stages
- Autotune & Autotrim indicators
- Modifier modes which might be combined with the self-contained ones (e.g. AltHold)

Potential issues fixed by this include: #1959, #1993 and #2032

fiam committed:
Add isArmingDisabledReason() function

Returns the 1st condition which is preventing arming

fiam committed:
Implement new OSD element OSD_MESSAGES

Used to display reasons for failure to arm

fiam committed:
Show the exact sensor which failed on HW failures

Kudos to @giacomo892 for the idea

fiam committed:
Wrap strings for messages with size-checking macro

Allows checking that message length doesn't exceed the maximum
message size at compile time.

fiam committed:
Add CLI setting for OSD_MESSAGES

digitalentity commented on src/main/fc/runtime_config.c:
```diff
@@ -29,6 +32,37 @@ uint32_t flightModeFlags = 0;
 
 static uint32_t enabledSensors = 0;
 
+// XXX: This function assumes that all flags in ARMING_DISABLED_ALL_FLAGS
```
Why? You can simply iterate through set bits of `ARMING_DISABLED_ALL_FLAGS` and check if corresponding bit it set in `armingFlags`.  You are returning the bitmask anyway.

I also think we should keep the code portable - avoid ARM assembly where possible.

fiam replied:
I tried to save a few cycles by checking just the relevant bits, but I will change that. Performance doesn't matter in this function since it will never be called while the aircraft is armed.

digitalentity commented on src/main/io/osd.c:
```diff
@@ -255,6 +268,89 @@ static void osdFormatCoordinate(char *buff, char sym, int32_t val)
+            return OSD_MESSAGE_STR("THROTTLE IS NOT LOW");
+        case ARMING_DISABLED_CLI:
+            return OSD_MESSAGE_STR("CLI IS ACTIVE");
+        case ARMING_DISABLED_CMS_MENU:
```
This would probably be never shown.

fiam replied:
Roger that! I wasn't quite sure of all the code paths that might set those bits. I'll remove the messages for `ARMING_DISABLED_CMS_MENU ` and `ARMING_DISABLED_OSD_MENU`

digitalentity commented on src/main/io/osd.c:
```diff
@@ -255,6 +268,89 @@ static void osdFormatCoordinate(char *buff, char sym, int32_t val)
+        case ARMING_DISABLED_CMS_MENU:
+            return OSD_MESSAGE_STR("CMS MODE IS ACTIVE");
+        case ARMING_DISABLED_OSD_MENU:
+            return OSD_MESSAGE_STR("OSD MENU IS ACTIVE");
```
This would probably be never shown as well

digitalentity commented on src/main/io/osd.c:
```diff
@@ -640,6 +736,25 @@ static bool osdDrawSingleElement(uint8_t item)
+            const char *message = NULL;
+            if (ARMING_FLAG(ARMED)) {
+                // TODO: Add some messages here
+            } else if (IS_RC_MODE_ACTIVE(BOXARM)) {
```
We may be trying to arm via sticks. We need to show the reason for blocked arming regardless of arming switch

fiam replied:
Hadn't thought of that. I'll add an `isTryingToArm()` function or something named like that to rc_modes.h and I'll use in osd.c.

digitalentity reviewed (commented):
Some things to consider, but in general - looks good.

fiam committed:
Remove messages indicating arming failure for impossible cases

Removed the messages for ARMING_DISABLED_CMS_MENU and
RMING_DISABLED_OSD_MENU

fiam committed:
Avoid using ASM in isArmingDisabledReason()

Makes the code easier to understand

fiam committed:
Always show a message when arming is not possible

Don't wait for the arming switch to go up, show the message at
all times while arming is not possible. Also, don't blink since
the switching between the "UNABLE TO ARM" and the actual message
is attention grabbing enough.

fiam committed:
Remove comment regarding OSD_MESSAGES default visibility

It's been decided to leave it on as a default
